flintrock.py

```diff
@@ -526,42 +526,104 @@ def launch_ec2(
         vpc_id=vpc_id,
         region=region)
 
+    machine_num = num_slaves + 1
     try:
         connection = boto.ec2.connect_to_region(region_name=region)
 
-        print("Launching {c} instances...".format(c=num_slaves + 1))
-
-        reservation = connection.run_instances(
-            image_id=ami,
-            min_count=(num_slaves + 1),
-            max_count=(num_slaves + 1),
-            key_name=key_name,
-            instance_type=instance_type,
-            placement=availability_zone,
-            security_group_ids=[sg.id for sg in security_groups],
-            subnet_id=subnet_id,
-            placement_group=placement_group,
-            tenancy=tenancy,
-            ebs_optimized=ebs_optimized,
-            instance_initiated_shutdown_behavior=instance_initiated_shutdown_behavior)
-
-        time.sleep(10)  # AWS metadata eventual consistency tax.
-
-        # TODO: Move this to a reusable function and add a limit on
-        #       wait time.
-        while True:
-            for instance in reservation.instances:
-                if instance.state == 'running':
-                    continue
+        #spot instance request
+        if spot_price is not None:
+            print("Requesting {c} spot instances at a max price of {p}".format(c=machine_num, p=spot_price))
+            
+            instance_reqs = connection.request_spot_instances(
+                price=spot_price,
+                image_id=ami,
+                count=(machine_num), 
+                key_name=key_name, 
+                instance_type=instance_type, 
+                #launch_group="launch-group-%s" % cluster_name,
+                placement=availability_zone,
+                security_group_ids=[sg.id for sg in security_groups],
+                subnet_id=subnet_id,
+                placement_group=placement_group,
+                ebs_optimized=ebs_optimized,
+                
+                )
+
+            spot_instance_request_ids = [sir.id for sir in instance_reqs]
+            # now we loop:
+            print("Waiting for spot instances to be granted...")
+            try:
+                while True:
```

Comment:
Is there any way to simplify this block of code? It's a bit hard to follow. Can we reuse, for example, [this pattern](https://github.com/nchammas/flintrock/blob/a1a5a7e8216b61eeaf8f33512c3da2b5142c542d/flintrock.py#L559-L568) for waiting on resources to reach a certain state?